packages/core-snapshots/src/codecs/json-codec.ts

```diff
@@ -0,0 +1,113 @@
+import { Container } from "@arkecosystem/core-kernel";
+import { camelizeKeys } from "xcase";
+
+import { Codec } from "../contracts";
+import { Codec as CodecException } from "../exceptions";
+
+@Container.injectable()
+export class JSONCodec implements Codec {
+    private static prepareData(data: any) {
+        if (Buffer.isBuffer(data)) {
+            return data.toJSON();
+        }
+
+        if (typeof data === "bigint") {
+            return data.toString();
+        }
+
+        return data;
+    }
+
+    private static removePrefix(item: Record<string, any>, prefix: string): any {
+        const itemToReturn = {};
+
+        for (const key of Object.keys(item)) {
+            itemToReturn[key.replace(prefix, "")] = JSONCodec.prepareData(item[key]);
+        }
+
+        return itemToReturn;
+    }
+
+    private static stringify(item: any): string {
+        return JSON.stringify(item, (_, v) => (typeof v === "bigint" ? `${v}n` : v));
+    }
+
+    private static parse(text: string) {
+        return JSON.parse(text, (_, value) => {
+            if (typeof value === "string") {
+                const m = value.match(/(-?\d+)n/);
+
+                if (m && m[0] === value) {
+                    value = BigInt(m[1]);
+                }
+            }
+
+            return value;
+        });
+    }
+
+    public blocksEncode(block): Buffer {
+        try {
+            const blockStringified = JSONCodec.stringify(camelizeKeys(JSONCodec.removePrefix(block, "Block_")));
+
+            return Buffer.from(blockStringified);
+        } catch (e) {
+            throw new CodecException.BlockEncodeException(block.Block_id);
+        }
+    }
+
+    public blocksDecode(buffer: Buffer): any {
+        try {
+            return JSON.parse(buffer.toString());
+        } catch (e) {
+            throw new CodecException.BlockDecodeException();
```

Comment:
Always include the original error object or error message when throwing a custom exception about something that involves data.